### PR TITLE
Support python 3.10

### PR DESCRIPTION
**Pull request by bjlittle**
### Description
This PR extends our CI `tests` GHA workflow coverage for `py310`.

This means we'll finally have our rolling active test coverage of `py38`, `py39` and `py310`.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by bjlittle** [comment]
There's an outage on GHA at the moment, and as a result the CI workflows haven't run for this PR.

I'll move it to `draft` and wait for the GH services to resume

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1345,7 +1345,15 @@ def __add__(self, mod):
...
-        return super().__hash__()
+        # Required in py310 due to change in behaviour for point=np.nan,
+        # as calling super().__hash__() returns a different hash and thus
+        # does not trigger the following call to __eq__ to determine equality.
+        # This is to align with pre-py310 behaviour.
+        try:
+            point = "nan" if np.isnan(self.point) else self.point
+        except TypeError:
+            point = self.point
+        return hash((point,))
```
Comment (bjlittle) [inline lib/iris/coords.py:1356]: Oh. My. Goodness.

[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/coords.py:1356]: Without this fix `iris.tests.integration.merge.TestNaNs.test_merge_nan_coords` fails, as comparing `Cell(point=nan, bound=None)` to an equivalent instance is actually deemed **not** equivalent. 

This subsequently results in a what should be a scalar coordinate in the resultant cube being promoted to a vector coordinate to cover a new dimension with duplicate `nan` point values.

Very obscure behaviour.

[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/coords.py]: @jamesp Updated the comments, see 4d47959

Thanks :+1:

[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/coords.py]: > scream Is there a performance hit here? How often is `__hash__` called?

For the failed integration test, I've compared the call sequence for `Cell.__hash__` and `Cell.__eq__` between `py38`+`np1.22.4` and `py310`+`np1.23.0`.

The fix in this PR results in the same call sequence i.e., we want to minimise the number of calls to `Cell.__eq__`, and keep `Cell.__hash__` as light-weight as possible; hence no explicit `Cell.bound` checks for `NaN`, it's not necessary as we can delegate that to `Cell.__eq__`.

My rudimentary timings of the failed integration test and running other merge unittests are comparable i.e., I can't see a real difference in performance :+1: 

TBH our benchmarking will be the judge to that :wink:

[the remark this answers is not in the document]
Comment (bjlittle) [reply lib/iris/coords.py]: That said, this change does make me feel...
image

**Review by jamesp — APPROVED** [review]

**Comment by lbdreyer** [comment]
It looks like the latest docs build failed with py 3.10: https://readthedocs.org/projects/scitools-iris/builds/17321543/

**Comment by bjlittle** [comment]
> > It looks like the latest docs build failed with py 3.10: https://readthedocs.org/projects/scitools-iris/builds/17321543/
> 
> That error looks familiar 🤔 [SciTools/cf-units#258](https://github.com/SciTools/cf-units/issues/258)

Essentially, for `py310` on rtd we need a release of `cf-units`... which has been ready for months. It seems to be doing a `pip install -e .` under the hood to install the `iris` src, and is then attempting to check and install all the package deps in the mamba/conda environment using PyPI. 

I may be able to work around it, by injecting a `--no-deps`, if possible. Worst case I'll release `cf-units` myself

File: lib/iris/coords.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/coords.py]: #4843 😔

**Pull request merged**